File: R/test_bridge.R
# Test R-Python Bridge via reticulate
# This script verifies that R can successfully call Python functions

# Load utilities
source("R/utils.R")

#' Run Bridge Tests
#'
#' Executes a series of tests to verify R-Python communication.
#'
#' @return Logical indicating if all tests passed
run_bridge_tests <- function() {
  cat("=== R-Python Bridge Test ===\n\n")

  # Initialize Python environment
  cat("1. Initializing Python environment...\n")
  init_python_env("dsr-portfolio")

  # Check Python config
  cat("\n2. Checking Python configuration...\n")
  config <- get_python_config()
  cat(paste("   Python:", config$python, "\n"))
  cat(paste("   Version:", config$version, "\n"))
  cat(paste("   NumPy available:", config$numpy, "\n"))
  cat(paste("   Pandas available:", config$pandas, "\n"))
  cat(paste("   PyTorch available:", config$torch, "\n"))

  # Source the test module
  cat("\n3. Sourcing Python test module...\n")
  source_python("python/test_bridge.py")

  # Test 1: Get equal weights
  cat("\n4. Testing get_test_weights()...\n")
  weights <- get_test_weights(5L)
  cat(paste("   Weights:", paste(round(weights, 4), collapse = ", "), "\n"))
  cat(paste("   Sum:", sum(weights), "\n"))

  test1_pass <- abs(sum(weights) - 1.0) < 1e-6
  cat(paste("   Test 1 PASSED:", test1_pass, "\n"))

  # Test 2: Get random weights
  cat("\n5. Testing get_random_weights()...\n")
  random_weights <- get_random_weights(5L, 42L)
  cat(paste("   Weights:", paste(round(random_weights, 4), collapse = ", "), "\n"))
  cat(paste("   Sum:", sum(random_weights), "\n"))

  test2_pass <- abs(sum(random_weights) - 1.0) < 1e-6
  cat(paste("   Test 2 PASSED:", test2_pass, "\n"))

  # Test 3: Array addition
  cat("\n6. Testing add_arrays()...\n")
  a <- c(1, 2, 3)
  b <- c(4, 5, 6)
  result <- add_arrays(a, b)
  expected <- c(5, 7, 9)
  cat(paste("   Input a:", paste(a, collapse = ", "), "\n"))
  cat(paste("   Input b:", paste(b, collapse = ", "), "\n"))
  cat(paste("   Result:", paste(result, collapse = ", "), "\n"))

  test3_pass <- all(result == expected)
  cat(paste("   Test 3 PASSED:", test3_pass, "\n"))

  # Test 4: Bridge info
  cat("\n7. Testing bridge_info()...\n")
  info <- bridge_info()
  cat(paste("   Python version:", info$python_version, "\n"))
  cat(paste("   NumPy version:", info$numpy_version, "\n"))

  test4_pass <- !is.null(info$python_version) && !is.null(info$numpy_version)
  cat(paste("   Test 4 PASSED:", test4_pass, "\n"))

  # Summary
  all_pass <- all(c(test1_pass, test2_pass, test3_pass, test4_pass))
  cat("\n=== Test Summary ===\n")
  cat(paste("Tests passed: ", sum(c(test1_pass, test2_pass, test3_pass, test4_pass)), "/4\n"))

  if (all_pass) {
    cat("\nSUCCESS: R-Python bridge is working correctly!\n")
  } else {
    cat("\nFAILURE: Some tests failed. Check the output above.\n")
  }

  return(all_pass)
}

# Run tests if executed directly
if (!interactive()) {
  success <- run_bridge_tests()
  quit(status = if (success) 0 else 1)
}

File: python/test_bridge.py
"""
Test module for R-Python bridge verification.
This module provides simple functions to verify reticulate can call Python.
"""

import numpy as np


def get_test_weights(n_assets: int) -> np.ndarray:
    """
    Return dummy equal weights for testing the R-Python bridge.

    Args:
        n_assets: Number of assets in portfolio

    Returns:
        NumPy array of equal weights summing to 1
    """
    return np.ones(n_assets) / n_assets


def get_random_weights(n_assets: int, seed: int = 42) -> np.ndarray:
    """
    Return random weights for testing.

    Args:
        n_assets: Number of assets in portfolio
        seed: Random seed for reproducibility

    Returns:
        NumPy array of random weights summing to 1
    """
    np.random.seed(seed)
    weights = np.random.random(n_assets)
    return weights / weights.sum()


def add_arrays(a: np.ndarray, b: np.ndarray) -> np.ndarray:
    """
    Add two arrays element-wise.

    Args:
        a: First array
        b: Second array

    Returns:
        Element-wise sum
    """
    return np.array(a) + np.array(b)


def bridge_info() -> dict:
    """
    Return information about the Python environment for debugging.

    Returns:
        Dictionary with Python version and numpy version
    """
    import sys
    return {
        "python_version": sys.version,
        "numpy_version": np.__version__
    }


if __name__ == "__main__":
    # Test functions locally
    print("Testing bridge functions...")

    weights = get_test_weights(5)
    print(f"Equal weights (5 assets): {weights}")
    print(f"Sum: {weights.sum()}")

    random_weights = get_random_weights(5)
    print(f"Random weights: {random_weights}")
    print(f"Sum: {random_weights.sum()}")

    a = np.array([1, 2, 3])
    b = np.array([4, 5, 6])
    print(f"Array addition: {a} + {b} = {add_arrays(a, b)}")

    print(f"Bridge info: {bridge_info()}")
